### PR TITLE
executor: make memory tracing in ProjectionExec more accurate

### LINKED ISSUE #16697: Projection Executor uses too much memory.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?

```mysql
CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL,
  `b` int(11) DEFAULT NULL
);

mysql> truncate table t;
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (1, 1);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t select * from t;
Query OK, 1 row affected (0.00 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> insert into t select * from t;
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> insert into t select * from t;
Query OK, 4 rows affected (0.00 sec)
Records: 4  Duplicates: 0  Warnings: 0

mysql> insert into t select * from t;
Query OK, 8 rows affected (0.00 sec)
Records: 8  Duplicates: 0  Warnings: 0

mysql> insert into t select * from t;
Query OK, 16 rows affected (0.00 sec)
Records: 16  Duplicates: 0  Warnings: 0

mysql> insert into t select * from t;
Query OK, 32 rows affected (0.00 sec)
Records: 32  Duplicates: 0  Warnings: 0

mysql> insert into t select * from t;
Query OK, 64 rows affected (0.00 sec)
Records: 64  Duplicates: 0  Warnings: 0

mysql> insert into t select * from t;
Query OK, 128 rows affected (0.00 sec)
Records: 128  Duplicates: 0  Warnings: 0

mysql> insert into t select * from t;
Query OK, 256 rows affected (0.00 sec)
Records: 256  Duplicates: 0  Warnings: 0

mysql> insert into t select * from t;
Query OK, 512 rows affected (0.01 sec)
Records: 512  Duplicates: 0  Warnings: 0
```

```mysql
mysql> explain analyze  select t1.a, t1.a +1 from t t1 join t t2 join t t3 order by t1.a;
+-----------------------------------+------------------+-----------+-----------+---------------+------------------------------------------------------------------------------------+----------------------------------------+-----------------------+---------+
| id                                | estRows          | actRows   | task      | access object | execution info                                                                     | operator info                          | memory                | disk    |
+-----------------------------------+------------------+-----------+-----------+---------------+------------------------------------------------------------------------------------+----------------------------------------+-----------------------+---------+
| Sort_10                           | 1000000000000.00 | 134217728 | root      |               | time:12.713145506s, loops:131073                                                   | test.t.a:asc                           | 3.21484375 GB         | 0 Bytes |
| └─Projection_12                   | 1000000000000.00 | 134217728 | root      |               | time:2.998321566s, loops:131073, Concurrency:4                                     | test.t.a, plus(test.t.a, 1)->Column#10 | 1.1401567347347736 GB | N/A     |
|   └─HashJoin_13                   | 1000000000000.00 | 134217728 | root      |               | time:2.814318715s, loops:131073, Concurrency:5, probe collision:0, build:105.04µs  | CARTESIAN inner join                   | 8.234375 KB           | 0 Bytes |
|     ├─TableReader_22(Build)       | 10000.00         | 512       | root      |               | time:5.177027ms, loops:2, rpc num: 1, rpc time:5.183946ms, proc keys:0             | data:TableFullScan_21                  | 4.2421875 KB          | N/A     |
|     │ └─TableFullScan_21          | 10000.00         | 512       | cop[tikv] | table:t3      | time:4.935532ms, loops:513                                                         | keep order:false, stats:pseudo         | N/A                   | N/A     |
|     └─HashJoin_15(Probe)          | 100000000.00     | 262144    | root      |               | time:7.162505ms, loops:257, Concurrency:5, probe collision:0, build:101.022µs      | CARTESIAN inner join                   | 8.234375 KB           | 0 Bytes |
|       ├─TableReader_20(Build)     | 10000.00         | 512       | root      |               | time:4.965126ms, loops:2, rpc num: 1, rpc time:5.186233ms, proc keys:0             | data:TableFullScan_19                  | 4.2421875 KB          | N/A     |
|       │ └─TableFullScan_19        | 10000.00         | 512       | cop[tikv] | table:t2      | time:4.940069ms, loops:513                                                         | keep order:false, stats:pseudo         | N/A                   | N/A     |
|       └─TableReader_18(Probe)     | 10000.00         | 512       | root      |               | time:4.619251ms, loops:2, rpc num: 1, rpc time:4.714003ms, proc keys:0             | data:TableFullScan_17                  | 4.2421875 KB          | N/A     |
|         └─TableFullScan_17        | 10000.00         | 512       | cop[tikv] | table:t1      | time:4.459083ms, loops:513                                                         | keep order:false, stats:pseudo         | N/A                   | N/A     |
+-----------------------------------+------------------+-----------+-----------+---------------+------------------------------------------------------------------------------------+----------------------------------------+-----------------------+---------+
10 rows in set (12.73 sec)
```
### 2. What did you expect to see?

The memory usage of Projection Executor is reasonable.

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)

master (a9091029b) , and run `./tidb-server` directly 

source: https://asktug.com/t/topic/33681

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #16697 

Problem Summary: make memory tracing in ProjectionExec more accurate

### What is changed and how it works?
`projectionWorker` may change the size of its input data chunk, but we don't trace this memory change.

### Check List 

Tests 

- Unit test

### Release note 
- Make memory tracing in `ProjectionExec` more accurate

### CODE REVIEW

**Comment by SunRunAway** [comment]
How's the `explain analyze` results in #16697 after this PR?

**Comment by qw4990** [comment]
> How's the `explain analyze` results in #16697 after this PR?

40~50 KB after this PR.

File: expression/integration_test.go [hunk 1]
```diff
@@ -6379,6 +6379,26 @@ func (s *testIntegrationSuite) TestIssue16779(c *C) {
...
+func (s *testIntegrationSuite) TestIssue16697(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("CREATE TABLE `t` (`a` int(11) DEFAULT NULL,`b` int(11) DEFAULT NULL)")
+	tk.MustExec("insert into t values (1, 1)")
+	for i := 0; i < 8; i++ {
+		tk.MustExec("insert into t select * from t")
+	}
+	rows := tk.MustQuery("explain analyze  select t1.a, t1.a +1 from t t1 join t t2 join t t3 order by t1.a").Rows()
```
Comment (SunRunAway) [inline expression/integration_test.go:6391]: nit: This case may run a very long time.

**Pull request merged**